client/components/data/query/index.jsx

```diff
@@ -0,0 +1,59 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import PropTypes from 'prop-types';
+import { PureComponent } from 'react';
+import { bindActionCreators } from 'redux';
+import { connect } from 'react-redux';
+import { values } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import compareProps from 'lib/compare-props';
+import { getRequest } from 'state/selectors';
+
+class Query extends PureComponent {
+	static propTypes = {
+		compareOptions: PropTypes.object,
+		requestAction: PropTypes.func.isRequired,
+		requesting: PropTypes.bool.isRequired,
+	};
+
+	componentWillMount() {
+		this.comparator = compareProps( {
+			ignore: [ 'compareOptions', 'requestAction', 'requesting' ],
+			...this.props.compareOptions,
+		} );
+		this.request( this.props );
+	}
+
+	componentWillReceiveProps( nextProps ) {
+		if ( ! this.comparator( this.props, nextProps ) ) {
+			this.request( nextProps );
+		}
+	}
```

Comment:
`this.comparator` should also be updated here if the `compareOptions` prop changes.

To ensure compatibility with `getDerivedStateFromProps`, could we store the `comparator` in `this.state`?